### PR TITLE
HHRE-344 Added new flows to crawl websites and download files

### PR DESCRIPTION
**Pull request by Harisai-Veeramallu**

### CODE REVIEW

File: spark_pipeline_framework/pipelines/framework_file_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                    event_text=f"Exception in Stage={stage_name}",
+                    ex=e,
+                )
+                raise friendly_spark_exception from e
+        self.progress_logger.log_event(
+            event_name=pipeline_name, event_text=f"Finished Pipeline {pipeline_name}"
+        )
+        return response
+
+    def transform(self, dataset, response, params=None):
```
Comment (changsu-bWell) [inline spark_pipeline_framework/pipelines/framework_file_pipeline.py:114]: I believe this method, `transform` in this file is redundant, as the transform method call in line 75 is actually defined in line 25-26 in the `FrameworkFileTransformer`.

File: spark_pipeline_framework/transformers/framework_file_downloader/v1/framework_file_downloader.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    # noinspection PyUnusedLocal
+    @keyword_only
+    def __init__(
+        self,
+        download_urls: List[str] = None,
+        download_to_path: Optional[str] = "./",
+        name: Optional[str] = None,
+        parameters: Optional[Dict[str, Any]] = None,
+        progress_logger: Optional[ProgressLogger] = None,
+        extract_zips: bool = False,
```
Comment (changsu-bWell) [inline spark_pipeline_framework/transformers/framework_file_downloader/v1/framework_file_downloader.py:27]: can we make this `extract_zips` variable Optional as well?
cos the file type we're trying to download may not be in an "archive" file format.

**Comment by changsu-bWell** [comment]
Just reviewed, a few points,
1. Now I understand better why you took a route to create another type of Transformer, `FrameworkPipeline` to meet the needs for this purpose - I believe this is a good approach, but lets's invite Imran for his buy-in for this approach, once it is ready,

File: spark_pipeline_framework/pipelines/framework_file_pipeline.py (hunk 1 above)
Comment (Harisai-Veeramallu) [reply spark_pipeline_framework/pipelines/framework_file_pipeline.py]: `transform` internally calls the `_transform` function and is not redundant.

File: library/pipelines/reference_data_download_test_pipeline/v1/reference_data_download_test_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+
+class ReferenceDataDownloadTestPipeline(FrameworkParamPipeline):
+    def __init__(self, parameters: Dict[str, Any], progress_logger: ProgressLogger):
+        super(ReferenceDataDownloadTestPipeline, self).__init__(
+            parameters=parameters, progress_logger=progress_logger
+        )
+
+        download_path: str = os.path.join(
+            os.path.dirname(os.path.realpath(__file__)), "data"
```
[the remark this answers is not in the document]
Comment (amin010) [reply library/pipelines/reference_data_download_test_pipeline/v1/reference_data_download_test_pipeline.py:27]: somwthing like: `Path(__file__).parent.join("data")`

File: library/pipelines/reference_data_download_test_pipeline/v1/reference_data_download_test_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        )
+
+        # df, response
+        # df
+        # "key", "value"
```
Comment (amin010) [inline library/pipelines/reference_data_download_test_pipeline/v1/reference_data_download_test_pipeline.py:32]: not sure if these commented likes are necessary

File: spark_pipeline_framework/transformers/framework_file_downloader/v1/framework_file_downloader.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    @keyword_only
+    def __init__(
+        self,
+        download_urls: Optional[List[str]] = [],
+        download_to_path: Optional[str] = "./",
+        param_key: Optional[str] = "urls",
+        name: Optional[str] = None,
+        parameters: Optional[Dict[str, Any]] = None,
+        progress_logger: Optional[ProgressLogger] = None,
+        extract_zips: Optional[bool] = False,
```
Comment (amin010) [inline spark_pipeline_framework/transformers/framework_file_downloader/v1/framework_file_downloader.py:28]: not a bad idea to document arguments

File: spark_pipeline_framework/utilities/file_downloader.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    @staticmethod
+    def get_filename_from_url(url: str) -> str:
+        filename = os.path.basename(url_parse.urlparse(url).path)
+        filename = filename.strip(" \n\t.")
+        if len(filename) == 0:
+            return None  # type: ignore
+        return filename
+
+    # @staticmethod
+    # def get_filename_from_headers(headers: Any) -> str:
```
Comment (amin010) [inline spark_pipeline_framework/utilities/file_downloader.py:44]: please specify why this method is commented

File: spark_pipeline_framework/utilities/file_downloader.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    #     if len(fnames) > 1:
+    #         return None
+    #     name = fnames[0].split('=')[1].strip(' \t"')
+    #     name = os.path.basename(name)
+    #     if not name:
+    #         return None
+    #     return name
+
+    @staticmethod
+    def rename_filename_if_exists(filename: str) -> str:
```
Comment (amin010) [inline spark_pipeline_framework/utilities/file_downloader.py:72]: I suggest GUID based method, this looks unnecessarily complicated

File: spark_pipeline_framework/utilities/file_downloader.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+
+class FileDownloader:
+    """
+    Module to download files from urls
+    Inspired from python wget module: https://github.com/steveeJ/python-wget
+    """
+
+    def __init__(
+        self, url: str, download_path: Optional[str] = "",
```
Comment (amin010) [inline spark_pipeline_framework/utilities/file_downloader.py:24]: I think it's likely that the `download_path` be an s3 url

Comment (Harisai-Veeramallu) [reply spark_pipeline_framework/utilities/file_downloader.py:24]: Ideally that should be the case, but some of the reference data files are obtained in the form of archive files. So the utility should be able to download the archive files onto the server, extract them and then upload them to the S3.
So the `download_path` in such cases would be a temporary location on the server and s3 url in other cases.

File: spark_pipeline_framework/utilities/file_downloader.py (hunk 7 above)
Comment (Harisai-Veeramallu) [reply spark_pipeline_framework/utilities/file_downloader.py:72]: GUID is a great way of downloading files, but I would be needing some extra effort to track the name mapping of multiple files as I need them properly named before uploading them to the s3 datalake.

File: spark_pipeline_framework/transformers/framework_file_downloader/v1/framework_file_downloader.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+        for url in urls:
+            with ProgressLogMetric(name=name or url, progress_logger=progress_logger):
+                downloader = FileDownloader(url=url, download_path=download_path)
+                result = downloader.download_files_from_url()
+                if extract_zips:
+                    extracted_result = downloader.extract_zip_files(
+                        result, download_path
+                    )
+                    result = extracted_result or result
```
Comment (amin010) [inline spark_pipeline_framework/transformers/framework_file_downloader/v1/framework_file_downloader.py:78]: maybe `result = FileDownloader(url=url, download_path=download_path, extract_zips)`?

Comment (amin010) [reply spark_pipeline_framework/transformers/framework_file_downloader/v1/framework_file_downloader.py:78]: not sure why we need `if extract_zips` logic out of FileDownloader class

File: spark_pipeline_framework/utilities/file_downloader.py (hunk 8 above)
Comment (amin010) [reply spark_pipeline_framework/utilities/file_downloader.py:24]: I was under the impression that FileDownloader handle unzipping automatically but still, the final destination is very likely s3 which doesn't seem we're handling it in FileDownloader class or anywhere else

**Review by imranq2 — APPROVED** [review]

**Review by changsu-bWell — APPROVED** [review]

File: spark_pipeline_framework/utilities/file_downloader.py [hunk 10]
```diff
@@ -81,13 +83,8 @@ def extract_zip_files(
 
         return None  # type: ignore
```
Comment (amin010) [inline spark_pipeline_framework/utilities/file_downloader.py:84]: it just needs an -> Optional[str]

**Pull request merged**